File: csharp/Day01/Day01.cs
namespace AdventOfCode2024.Day01;

public static class Day01
{
    public static int PartOne(string[] fileLines)
    {
        var (left, right) = ParseFile(fileLines);
        
        return left
            .Order()
            .Zip(right.Order().ToArray(), (l, r) => Math.Abs(l - r))
            .Sum();
    }
    
    public static int PartTwo(string[] fileLines)
    {
        var (left, right) = ParseFile(fileLines);
        return left.Select(l => l * right.Count(r => r == l)).Sum();
    }

    private static (int[] Left, int[] Right) ParseFile(string[] lines)
    {
        var (left, right) = (new int[lines.Length], new int[lines.Length]);

        for (var i = 0; i < lines.Length; i++)
        {
            var parts = lines[i].Split("   ");
            (left[i], right[i]) = (int.Parse(parts[0]), int.Parse(parts[1]));
        }

        return (left, right);
    }
}

File: fsharp/day01.fs
module AoC.fsharp.day01

let parseFile(lines: string array) =
    let left  = Array.zeroCreate lines.Length
    let right = Array.zeroCreate lines.Length
    
    lines |> Array.iteri(fun i l ->
        let parts = l.Split("   ")
        left[i]  <- int parts[0]
        right[i] <- int parts[1]
    )

    (left, right)
    
let part1 (lines: string array) =
    let left, right = (parseFile lines)
    Array.sortInPlace left
    Array.sortInPlace right
    left |> Array.zip right |> Array.sumBy(fun (l, r) -> abs (l - r))

let part2 (lines: string array) =
    let left, right = parseFile lines
    left |> Array.sumBy(fun l -> l * (right |> Array.filter(fun r -> r = l) |> Array.length))
